sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -514,6 +514,25 @@ case class OptimizeCodegen(conf: CatalystConf) extends Rule[LogicalPlan] {
 
 
 /**
+ * Reorders the predicates in `Filter` so more expensive expressions like UDF can evaluate later.
+ */
+object ReorderPredicatesInFilter extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Filter(pred, child) =>
+      // Reverses the expressions to get the suffix deterministic expressions in the predicate.
+      // E.g., the original expressions are 'a > 1, rand(0), 'b > 2, 'c > 3.
+      // The reversed expressions are 'c > 3, 'b > 2, rand(0), 'a > 1.
+      // The suffix deterministic expressions are 'c > 3, 'b > 2.
+      val (deterministicExprs, others) = splitConjunctivePredicates(pred).reverse
```

Comment:
the split is widely used in optimizer. i think i may rewrite this reverse and span to alleviate performance concern.